File: Algorithms/java/046.Permutations.java
public class Solution {
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> result = new ArrayList<List<Integer>>();
        List<Integer> permutation = new ArrayList<>();
        boolean[] isVisited = new boolean[nums.length];
        DFS(result, permutation, nums, isVisited);
        return result;
    }
    private void DFS(List<List<Integer>> result, List<Integer> permutation,
                        int[] nums, boolean[] isVisited) {
        if (permutation.size() == nums.length) {
            result.add(new ArrayList<>(permutation));
            return;
        }
        for (int i = 0; i < nums.length; i++) {
            if (isVisited[i] == true) continue;
            isVisited[i] = true;
            permutation.add(nums[i]);
            DFS(result, permutation, nums, isVisited);
            permutation.remove(permutation.size()-1);
            isVisited[i] = false;
        }
    }
    /*
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> result = new ArrayList<List<Integer>>();
        permute(nums, 0, result);
        return result;
    }
    public void permute(int[] nums, int index, List<List<Integer>> result) {
        if (nums.length == index) {
            List<Integer> temp = new ArrayList<>();
            for (int i = 0; i < nums.length; i++) {
                temp.add(nums[i]);
            }
            result.add(temp);
            return;
        }
        for (int i=index; i < nums.length; ++i) {
            int tmp = nums[i];
            nums[i] = nums[index];
            nums[index] = tmp;
            permute(nums, index+1, result);
            tmp = nums[i];
            nums[i] = nums[index];
            nums[index] = tmp;
        }
    }
    */
    /*
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> result = new ArrayList<List<Integer>>();
        List<Integer> temp = new ArrayList<>();
        if (nums.length == 1) {
            temp.add(nums[0]);
            result.add(temp);
            return result;
        }
        int[] sub = new int[nums.length-1];
        for (int i=0; i < sub.length; i++) {
            sub[i] = nums[i+1];
        }
        List<List<Integer>> v = permute(sub);
        for (int i=0; i < v.size(); ++i) {
            for (int j=0; j < v.get(i).size(); j++) {
                temp = new ArrayList<>(v.get(i));
                int tmp = temp.get(j);
                temp.set(j, nums[0]);
                temp.add(0, tmp);
                result.add(temp);
            }
            temp = new ArrayList<>(v.get(i));
            temp.add(0, nums[0]);
            result.add(temp);
        }
        return result;
    }
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> result = new ArrayList<List<Integer>>();
        List<Integer> temp = new ArrayList<>();
        if (nums.length == 1) {
            temp.add(nums[0]);
            result.add(temp);
            return result;
        }
        int[] sub = new int[nums.length-1];
        for (int i=0; i < sub.length; i++) {
            sub[i] = nums[i+1];
        }
        List<List<Integer>> v = permute(sub);
        for (int i=0; i < v.size(); ++i) {
            for (int j=0; j <= v.get(i).size(); j++) {
                temp = new ArrayList<>(v.get(i));
                temp.add(j, nums[0]);
                result.add(temp);
            }
        }
        return result;
    }
    */
}

File: Algorithms/cpp/046.Permutations.cpp
class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> result;
        vector<int> permutation;
        vector<bool> isVisited(nums.size(), false);
        DFS(result, permutation, nums, isVisited);
        return result;
    }
    void DFS(vector<vector<int>>& result, vector<int>& permutation, vector<int>& nums, vector<bool>& isVisited) {
        if (permutation.size() == nums.size()) {
            result.push_back(permutation);
            return;
        }
        for (int i = 0; i < nums.size(); i++) {
            if (isVisited[i] == true) continue;
            isVisited[i] = true;
            permutation.push_back(nums[i]);
            DFS(result, permutation, nums, isVisited);
            permutation.pop_back();
            isVisited[i] = false;
        }
    }
    /*
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> result;
        permute(nums, 0, result);
        return result;
    }
    void permute(vector<int>& nums, int index,
            vector<vector<int>>& result) {
        if (nums.size() == index) {
            result.push_back(nums);
            return;
        }
        for (int i=index; i < nums.size(); ++i) {
            int temp = nums[i];
            nums[i] = nums[index];
            nums[index] = temp;
            permute(nums, index+1, result);
            temp = nums[i];
            nums[i] = nums[index];
            nums[index] = temp;
        }
    }
    */
    /*
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> result;
        vector<int> temp;
        if (nums.size() == 1) {
            temp.push_back(nums[0]);
            result.push_back(temp);
            return result;
        }
        vector<int> sub(nums.begin()+1, nums.end());
        vector<vector<int>> v = permute(sub);
        for (int i=0; i < v.size(); ++i) {
            for (int j=0; j < v[i].size(); j++) {
                temp.clear();
                temp = v[i];
                int tmp = temp[j];
                temp[j] = nums[0];
                temp.insert(temp.begin(), tmp);
                result.push_back(temp);
            }
            temp = v[i];
            temp.insert(temp.begin(), nums[0]);
            result.push_back(temp);
        }
        return result;
    }
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> result;
        vector<int> temp;
        if (nums.size() == 1) {
            temp.push_back(nums[0]);
            result.push_back(temp);
            return result;
        }
        vector<int> sub(nums.begin()+1, nums.end());
        vector<vector<int>> v = permute(sub);
        for (int i=0; i < v.size(); ++i) {
            for (int j=0; j <= v[i].size(); j++) {
                temp = v[i];
                temp.insert(temp.begin()+j, nums[0]);
                result.push_back(temp);
            }
        }
        return result;
    }
    */
};
